cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -27,19 +32,138 @@
+
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), Lists.<NodeValueEntry>newArrayList());
+      }
+
+      // Iterate over the list of NodeValueEntry corresponding to the current key.
+      // Only add those NodeValueEntry to the merged token which already do not exist.
+      for (NodeValueEntry otherNodeValueEntry : otherTokenValueMap.get(entry.getKey())) {
+        boolean otherNodeValueEntryExist = false;
+        for (NodeValueEntry thisNodeValueEntry : tokenValueMap.get(entry.getKey())) {
+          if (thisNodeValueEntry.equals(otherNodeValueEntry)) {
+            otherNodeValueEntryExist = true;
+            break;
+          }
+        }
+        if (!otherNodeValueEntryExist) {
+          tokenValueMap.get(entry.getKey()).add(otherNodeValueEntry);
+        }
+      }
+    }
+    if (other.getMapReduceCounters() != null) {
+      setMapReduceCounters(other.getMapReduceCounters());
+    }
+  }
+
+  @Override
+  public void put(String key, String value) {
+    if (nodeName == null) {
+      throw new IllegalStateException("Node name cannot be null.");
+    }
+    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
+    if (nodeValueList == null) {
+      nodeValueList = Lists.newArrayList();
+      tokenValueMap.put(key, nodeValueList);
+    }
+    nodeValueList.add(new NodeValueEntry(nodeName, value));
+  }
+
+  @Override
+  public List<NodeValueEntry> getAll(String key) {
+    if (containsKey(key)) {
+      return ImmutableList.copyOf(tokenValueMap.get(key));
+    }
+    return ImmutableList.of();
+  }
+
+  @Override
+  public Map<String, String> getAllFromNode(String nodeName, @Nullable String prefix) {
+    Map<String, String> tokenValuesFromNode = Maps.newHashMap();
+    for (Map.Entry<String, List<NodeValueEntry>> entry : tokenValueMap.entrySet()) {
+
+      if (prefix != null) {
+        if (!entry.getKey().equals(prefix) && !entry.getKey().startsWith(prefix + ".")) {
+          continue;
+        }
+      }
+
+      List<NodeValueEntry> nodeValueEntryList = entry.getValue();
+      for (NodeValueEntry nodeValueEntry : nodeValueEntryList) {
+        if (nodeValueEntry.getNodeName().equals(nodeName)) {
+          tokenValuesFromNode.put(entry.getKey(), nodeValueEntry.getValue());
+        }
+      }
+    }
+    return tokenValuesFromNode;
+  }
+
+  @Nullable
+  @Override
+  public String get(String key, String nodeName) {
+    if (!containsKey(key)) {
+      return null;
+    }
+
+    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
+    for (NodeValueEntry nodeValue : nodeValueList) {
+      if (nodeValue.getNodeName().equals(nodeName)) {
+        return nodeValue.getValue();
+      }
+    }
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public String get(String key) {
+    if (!containsKey(key)) {
```

Comment:
you actually don't need this check. Just do a get from the map and it returns `null`, then return `null`.